src/v/cluster/producer_state_manager.cc

```diff
@@ -48,68 +44,44 @@ void producer_state_manager::setup_metrics() {
       prometheus_sanitize::metrics_name("cluster:producer_state_manager"),
       {sm::make_gauge(
         "producer_manager_total_active_producers",
-        [this] { return _num_producers; },
+        [this] { return _cache.get_stats().total_size; },
         sm::description(
           "Total number of active idempotent and transactional producers."))});
 }
 
-void producer_state_manager::register_producer(producer_state& state) {
-    link(state);
-    ++_num_producers;
-    vlog(clusterlog.debug, "Registered producer: {}", state);
-}
-
-void producer_state_manager::deregister_producer(producer_state& state) {
-    if (state._hook.is_linked()) {
-        state._hook.unlink();
-        --_num_producers;
-        vlog(clusterlog.debug, "Removing producer: {}", state);
-    }
+void producer_state_manager::register_producer(
+  producer_state& state, std::optional<vcluster_id> vcluster) {
+    vlog(
+      clusterlog.debug,
+      "Registering producer: {}, current producer count: {}",
+      state,
+      _cache.get_stats().total_size);
+    _cache.insert(vcluster.value_or(no_vcluster), state);
 }
 
-void producer_state_manager::link(producer_state& state) {
-    vassert(
-      !state._hook.is_linked(),
-      "double linking of producer state {}",
-      state._id);
-    _lru_producers.push_back(state);
+void producer_state_manager::deregister_producer(
+  producer_state& state, std::optional<vcluster_id> vcluster) {
+    vlog(
+      clusterlog.debug,
+      "Removing producer: {}, current producer count: {}",
+      state,
+      _cache.get_stats().total_size);
+    _cache.remove(vcluster.value_or(no_vcluster), state);
 }
-
-bool producer_state_manager::can_evict_producer(
-  const producer_state& state) const {
-    return _num_producers > _max_ids()
-           || state.ms_since_last_update() > _producer_expiration_ms;
+void producer_state_manager::touch(
+  producer_state& state, std::optional<vcluster_id> vcluster) {
+    vlog(clusterlog.trace, "Touched producer: {}", state);
+    _cache.touch(vcluster.value_or(no_vcluster), state);
 }
 
-void producer_state_manager::evict_excess_producers() {
```

Comment:
IIUC there is no periodic eviction fiber, so we evict only if a new insert request comes in and the cache happens to be full?

Reply:
yes, exactly we chatted about that previously, the cache is only evicted when full